vsphere/datadog_checks/vsphere/api.py

```diff
@@ -0,0 +1,174 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+import functools
+import ssl
+
+from pyVim import connect
+from pyVmomi import vim, vmodl
+
+from datadog_checks.base import ensure_unicode
+from datadog_checks.vsphere.constants import ALL_RESOURCES, MAX_QUERY_METRICS_OPTION, UNLIMITED_HIST_METRICS_PER_QUERY
+
+
+def smart_retry(f):
+    """A function decorated with this `@smart_retry` will trigger a new authentication if it fails. The function
+    will then be retried.
+    This is useful when the integration keeps a semi-healthy connection to the vSphere API"""
+
+    @functools.wraps(f)
+    def wrapper(api_instance, *args, **kwargs):
+        try:
+            return f(api_instance, *args, **kwargs)
+        except Exception as e:
+            api_instance.log.debug(
+                "An exception occurred when executing %s: %s. Refreshing the connection to vCenter and retrying",
+                f.__name__,
+                e,
+            )
+            api_instance.smart_connect()
+            return f(api_instance, *args, **kwargs)
+
+    return wrapper
+
+
+class APIConnectionError(Exception):
+    pass
+
+
+class VSphereAPI(object):
+    """Abstraction class over the vSphere SOAP api using the pyvmomi library"""
+
+    def __init__(self, config, log):
+        self.config = config
+        self.log = log
+
+        self._conn = None
+        self.smart_connect()
+
+    def smart_connect(self):
+        """Creates the connection object to the vSphere API using parameters supplied from the configuration.
+        """
+        context = None
+        if not self.config.ssl_verify:
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
```

Comment:
https://docs.python.org/3/library/ssl.html#ssl.PROTOCOL_SSLv23

Do https://github.com/DataDog/integrations-core/blob/ade080d1432fd836a92545256a0f629ae33e849d/tls/datadog_checks/tls/tls.py#L20

Reply:
Should not be used in this if block (as it will prevent to set `context.verify_mode = ssl.CERT_NONE`).

But added in the next if block

Reply:
See https://docs.python.org/3/library/ssl.html#ssl.SSLContext.check_hostname for reference

Reply:
Done